shared/middleware/files_api.rb

```diff
@@ -400,6 +400,18 @@ def put_file(endpoint, encrypted_channel_id, filename, body)
       return bad_request if share_failure && share_failure[:type] != "address"
     end
 
+    # Don't allow project to be saved if it contains non-UTF-8 characters (causing error / project to not load when opened).
+    if 'sources' == endpoint
+      body_json = JSON.parse(body)
+      source = body_json["source"]
+      html = body_json["html"]
+
+      source_is_valid = source && source.force_encoding("UTF-8").valid_encoding?
```

Comment:
@madelynkasula One of the test failures for Java lab revealed that in Java lab, the `source` object has some nested objects:

`{ "source": { "MyClass.java": "…code…" } }` 

Does this happen anywhere else that you know of? I'm wondering if we should check for nested objects and parse through the keys if this is the case?

Reply:
ahhh, great find. i _think_ javalab is the first/only place to use this structure. 

brief explanation behind the reason for that: javalab is technically multi-file (so the source contains the nested file sources, but they're all saved in a single `main.json`) but we didn't want it to actually have multiple source files the way weblab does because it creates lots of pain for us in weblab 😄 

so even though this is only currently true for javalab, we should support this use case because i think eventually weblab should transition to this same structure, and we may have more multi-file labs in the future.